### PR TITLE
Feature/208 4 w filter in entity list

### PR DESCRIPTION
**Pull request by pieroit**

### CODE REVIEW

File: src/admin/class-wordlift-admin-entity-list.php [hunk 1]
```diff
@@ -80,4 +80,73 @@ public function render_custom_columns( $column, $entity_id ) {
...
+	/**
+	 * Add 4W select box before the 'Filter' button.
+	 *
+	 * @since 3.3.0
+	 */
+	public function add_4W_filter() {
+		global $typenow;
+		
+		// Only show on entity list page
+		if( $typenow !== Wordlift_Entity_Service::TYPE_NAME ){
```
Comment (ziodave) [inline src/admin/class-wordlift-admin-entity-list.php]: There is method in Entity Service which checks whether a type is an entity. Better use that.

File: src/admin/class-wordlift-admin-entity-list.php [hunk 2]
```diff
@@ -80,4 +80,73 @@ public function render_custom_columns( $column, $entity_id ) {
...
+		
+		// Build select box with the 4W
+		$all_w = array(
+			'Select W',
+			WL_WHAT_RELATION,
+			WL_WHO_RELATION,
+			WL_WHERE_RELATION,
+			WL_WHEN_RELATION
+		);
+		$output  = '<select name="4W_filter" id="dropdown_4W_type">';
```
Comment (ziodave) [inline src/admin/class-wordlift-admin-entity-list.php]: Do not use underscores in html attributes (name/id), use dash instead: https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Underscores_in_class_and_ID_Names

File: src/admin/class-wordlift-admin-entity-list.php [hunk 3]
```diff
@@ -80,4 +80,73 @@ public function render_custom_columns( $column, $entity_id ) {
...
+		foreach ( $all_w as $w ) {
+			$default  = ( $selected == $w ) ? 'selected' : '';
```
Comment (ziodave) [inline src/admin/class-wordlift-admin-entity-list.php]: Make comparison strict: ===

File: src/admin/class-wordlift-admin-entity-list.php [hunk 4]
```diff
@@ -80,4 +80,73 @@ public function render_custom_columns( $column, $entity_id ) {
...
+		);
+		$output  = '<select name="4W_filter" id="dropdown_4W_type">';
+		foreach ( $all_w as $w ) {
+			$default  = ( $selected == $w ) ? 'selected' : ''; 
+			$output  .= "<option value='$w' $default >" . ucfirst( __( $w, 'wordlift' ) ) . '</option>';
+		}
+		$output .= '</select>';
+		
+		// Print on page
+		echo $output;
```
Comment (ziodave) [inline src/admin/class-wordlift-admin-entity-list.php]: if you're echoing the variable output, you can echo each fragment, no need to collect and then to echo at the end.

File: src/admin/class-wordlift-admin-entity-list.php [hunk 5]
```diff
@@ -80,4 +80,73 @@ public function render_custom_columns( $column, $entity_id ) {
...
+	 * @param array $clauses WP main query clauses.
+	 * 
+	 * @return array Modified clauses.
+	 */
+	public function add_4W_filter_query( $clauses ) {
+				
+		global $typenow, $wpdb;
+		
+		// Only apply on entity list page and if the 4W_filter query param is set
+		if( ! ( $typenow == Wordlift_Entity_Service::TYPE_NAME && isset( $_GET['4W_filter'] ) ) ) {
```
Comment (ziodave) [inline src/admin/class-wordlift-admin-entity-list.php]: Use the method in Entity Service to determine whether is an entity.

File: src/admin/class-wordlift-admin-entity-list.php [hunk 6]
```diff
@@ -80,4 +80,73 @@ public function render_custom_columns( $column, $entity_id ) {
...
+			return $clauses;
+		}
+			
+		// Check a valid W was requested
+		$requested_w = $_GET['4W_filter'];
+		if ( ! in_array( $requested_w, array( WL_WHAT_RELATION, WL_WHO_RELATION, WL_WHERE_RELATION, WL_WHEN_RELATION )) ) {
```
Comment (ziodave) [inline src/admin/class-wordlift-admin-entity-list.php]: Fix formatting

File: src/admin/class-wordlift-admin-entity-list.php [hunk 7]
```diff
@@ -80,4 +80,73 @@ public function render_custom_columns( $column, $entity_id ) {
...
+			return $clauses;
+		}
+		
+		$w_table = wl_core_get_relation_instances_table_name();
+		
+		// Change WP main query clauses
+		$clauses['join']     .= "INNER JOIN $w_table ON $wpdb->posts.ID = $w_table.object_id";
+		$clauses['where']    .= "AND $w_table.predicate = '$requested_w'";
```
Comment (ziodave) [inline src/admin/class-wordlift-admin-entity-list.php]: I think this needs to be sanitized?

File: src/includes/class-wordlift.php [hunk 8]
```diff
@@ -389,10 +389,14 @@ private function define_admin_hooks() {
...
 
-		// Add custom columns for entity listing in the backand
+		// Entity listing customization (wp-admin/edit.php)
+		// Add custom columns
 		$this->loader->add_filter( 'manage_entity_posts_columns', $this->entity_list_service, 'register_custom_columns' );
 		$this->loader->add_filter( 'manage_entity_posts_custom_column', $this->entity_list_service, 'render_custom_columns', 10, 2 );
-
+		// Add 4W selection
+		$this->loader->add_action( 'restrict_manage_posts', $this->entity_list_service, 'add_4W_filter' );
+		$this->loader->add_filter( 'posts_clauses', $this->entity_list_service, 'add_4W_filter_query' );
```
Comment (ziodave) [inline src/includes/class-wordlift.php]: I would name the methods after the actions/filters names to make it easy to understand the binding between the method and the action/filter.

File: src/admin/class-wordlift-admin-entity-list.php (hunk 1 above)
Comment (pieroit) [inline src/admin/class-wordlift-admin-entity-list.php]: the method in `Wordlift_Entity_Service` is for a single entity:
```
public function is_entity( $post_id ) {

		return ( self::TYPE_NAME === get_post_type( $post_id ) );
	}
```
in this case the check is on the post_type of all the elements in the page, so no specific ID.

File: src/admin/class-wordlift-admin-entity-list.php (hunk 7 above)
Comment (pieroit) [inline src/admin/class-wordlift-admin-entity-list.php]: I have difficulties using prepare:
```
WordPress database error: [You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near ''wp_wl_relation_instances' ON 'wp_posts'.ID = 'wp_wl_relation_instances'.object_' at line 1]
SELECT SQL_CALC_FOUND_ROWS DISTINCT wp_posts.ID FROM wp_posts INNER JOIN 'wp_wl_relation_instances' ON 'wp_posts'.ID = 'wp_wl_relation_instances'.object_id WHERE 1=1 AND wp_posts.post_type = 'entity' AND (wp_posts.post_status = 'publish' OR wp_posts.post_status = 'future' OR wp_posts.post_status = 'draft' OR wp_posts.post_status = 'pending' OR wp_posts.post_status = 'private')AND wp_wl_relation_instances.predicate = 'what' ORDER BY wp_posts.post_date DESC LIMIT 0, 20
```
Seems mandatory to be used with user supplied params, but not with table names:
http://wordpress.stackexchange.com/questions/93830/really-simple-query-giving-error-in-sql-syntax

File: src/admin/class-wordlift-admin-entity-list.php [hunk 9]
```diff
@@ -80,4 +80,81 @@ public function render_custom_columns( $column, $entity_id ) {
...
+	public function posts_clauses_classification_scope( $clauses ) {
+		
+		// Run only on admin page
+		if( ! is_admin() ) {
+			return $clauses;
+		}
+		
+		// Only apply on entity list page, only if this is the main query and if the wl-classification-scope query param is set
+		$screen = get_current_screen();
+		if( ! ( $screen->post_type === Wordlift_Entity_Service::TYPE_NAME && is_main_query() && isset( $_GET['wl-classification-scope'] ) ) ) {
```
Comment (mcolacino) [inline src/admin/class-wordlift-admin-entity-list.php]: Use *is_entity* method

File: src/admin/class-wordlift-admin-entity-list.php [hunk 10]
```diff
@@ -80,4 +80,81 @@ public function render_custom_columns( $column, $entity_id ) {
...
+		if( $screen->post_type !== Wordlift_Entity_Service::TYPE_NAME ){
+			return;
+		}
+		
+		// Was a W already selected?
+		$selected = isset( $_GET['wl-classification-scope'] ) ? $_GET['wl-classification-scope'] : '' ;
+		
+		// Print select box with the 4W
+		$all_w = array(
+			"All 'W'",
```
Comment (mcolacino) [inline src/admin/class-wordlift-admin-entity-list.php]: Use translations here

File: src/admin/class-wordlift-admin-entity-list.php [hunk 11]
```diff
@@ -80,4 +80,81 @@ public function render_custom_columns( $column, $entity_id ) {
...
+			WL_WHAT_RELATION,
+			WL_WHO_RELATION,
+			WL_WHERE_RELATION,
+			WL_WHEN_RELATION
+		);
+		echo '<select name="wl-classification-scope" id="wl-dropdown-classification-scope">';
+		foreach ( $all_w as $w ) {
+			$default  = ( $selected === $w ) ? 'selected' : '';
+			echo sprintf( '<option value="%s" %s >%s</option>', $w, $default, $w );
```
Comment (mcolacino) [inline src/admin/class-wordlift-admin-entity-list.php]: Use here the constant as a value and the coresponding translation as label

File: src/admin/class-wordlift-admin-entity-list.php (hunk 9 above)
Comment (pieroit) [inline src/admin/class-wordlift-admin-entity-list.php]: Cannot use `is_entity` because there is no `$entity_id`

Comment (mcolacino) [inline src/admin/class-wordlift-admin-entity-list.php]: Pass the Entity Service trough the constructor ... Then you have the method available

Comment (pieroit) [inline src/admin/class-wordlift-admin-entity-list.php]: The check is not on 1 entity, is on an entity list page!!! No entity ID is available there for the method to check.

**Pull request merged**